data-prepper-plugins/mutate-event-processors/src/main/java/org/opensearch/dataprepper/plugins/processor/mutateevent/MapToListProcessor.java

```diff
@@ -49,36 +52,75 @@ public Collection<Record<Event>> doExecute(final Collection<Record<Event>> recor
-                final List<Map<String, Object>> targetList = new ArrayList<>();
-
-                Map<String, Object> modifiedSourceMap = new HashMap<>();
-                for (final Map.Entry<String, Object> entry : sourceMap.entrySet()) {
-                    if (excludeKeySet.contains(entry.getKey())) {
-                        if (config.getRemoveProcessedFields()) {
-                            modifiedSourceMap.put(entry.getKey(), entry.getValue());
+                final Map<String, Object> sourceMap = getSourceMap(recordEvent);
+
+                if (config.getConvertFieldToList()) {
+                    final List<List<Object>> targetNestedList = new ArrayList<>();
+
+                    for (final Map.Entry<String, Object> entry : sourceMap.entrySet()) {
+                        if (!excludeKeySet.contains(entry.getKey())) {
+                            targetNestedList.add(List.of(entry.getKey(), entry.getValue()));
                         }
-                        continue;
-                    }
-                    targetList.add(Map.of(
-                            config.getKeyName(), entry.getKey(),
-                            config.getValueName(), entry.getValue()
-                    ));
-                }
 
-                if (config.getRemoveProcessedFields()) {
-                    recordEvent.put(config.getSource(), modifiedSourceMap);
+                    }
+                    removeProcessedFields(sourceMap, recordEvent);
+                    recordEvent.put(config.getTarget(), targetNestedList);
+                } else {
+                    final List<Map<String, Object>> targetList = new ArrayList<>();
+                    for (final Map.Entry<String, Object> entry : sourceMap.entrySet()) {
+                        if (!excludeKeySet.contains(entry.getKey())) {
+                            targetList.add(Map.of(
+                                    config.getKeyName(), entry.getKey(),
+                                    config.getValueName(), entry.getValue()
+                            ));
+                        }
+                    }
+                    removeProcessedFields(sourceMap, recordEvent);
+                    recordEvent.put(config.getTarget(), targetList);
                 }
-
-                recordEvent.put(config.getTarget(), targetList);
             } catch (Exception e) {
                 LOG.error("Fail to perform Map to List operation", e);
-                //TODO: add tagging on failure
+                recordEvent.getMetadata().addTags(config.getTagsOnFailure());
             }
         }
         return records;
     }
 
+    private Map<String, Object> getSourceMap(Event recordEvent) throws JsonProcessingException {
+        final Map<String, Object> sourceMap;
+        if (config.getSource() == null) {
+            // Source is root
+            sourceMap = OBJECT_MAPPER.treeToValue(recordEvent.getJsonNode(), Map.class);
```

Comment:
I have a high level question here, if you get the entire JSON node if source is null, I think it's not going to work. Correct me if I am wrong. 
For example, if the following is the event.
```
{
  "my-map": {
    "key1": "value1",
    "key2": "value2",
    "key3": "value3"
  }
}
```

The result will be
```
[
  ["my-map", {
    "key1": "value1",
    "key2": "value2",
    "key3": "value3"
  }]
]
```

Reply:
Good point. I think adding test cases and documentation for euch cases would help.

Reply:
@asifsmohammed Good question, though it isn't necessarily related to root being the source. I think it's an issue with nested map in the source map. Currently we assume the map structure is flat. We can have a recursive option to go deeper into each field in the future if needed.

Reply:
Added readme and more test cases.